### PR TITLE
Feat: Use `UptimeCalculator` for PingChart

### PR DESCRIPTION
**Pull request by chakflying**
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Depends on #4267
Depends on #4406
Resolves https://github.com/louislam/uptime-kuma/issues/2251 as the jitter is now "printed"

Update PingChart to use data from the new `UptimeCalculator`. 

- Use 3 line charts to show the min, max and avg ping together for better visualization
- Avoid modifying `heartbeatList`, and instead store the received data in a separate data structure.
- With this method, there seems to be no easy way to show the maintenance period. How should we handle this? Adding another column in the `stat_minutely` tables seems wasteful.

## Type of change

Please delete any options that are not relevant.

- User interface (UI)
- New feature (non-breaking change which adds functionality)
- Breaking change (a fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I ran ESLint and other linters for modified files
- [ ] I have performed a self-review of my own code and tested it
- [ ] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [ ] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

image

image

### CODE REVIEW

**Comment by chakflying** [comment]
Realized the `stat_daily` data is not really useful for the chart. There would only be 7 data points for the 1 week chart, and there would be no meaningful information about when up/down events occurred.

Ideally we'd have a `stat_hourly` covering a couple of weeks. Currently going from minute to day is a 2 orders of magnitude increase, and one might say it's too big anyway. But doing this would also increase a bit of complexity, and I don't know if the data would be useful elsewhere.

**Comment by chakflying** [comment]
Rebased on master, did a quick test and seems to work okay. Still have a couple of issues:

- Will return value 0 even for monitors without ping value. Might need special handling for this
- No handling for maintenance periods

**Comment by louislam** [comment]
> Depends on https://github.com/louislam/uptime-kuma/pull/4267

#4267 had been merged, for your info.

**Comment by chakflying** [comment]
@louislam have you thought about how to store past maintenance information? I'm leaning towards adding a new column in the stats tables, so that it's indexable. However, if we think about having many other values to keep track of in the future, it may be better to have an "extras" column, then json encode the vales. `{"maintenance": 10, "slow_response": 2}`

**Comment by louislam** [comment]
Yes, I think it is a good idea to have a column for extra stats.

**Comment by chakflying** [comment]
Maintenance should now also be shown.

image

Branch will be rebased again once #4406 is merged.

**Comment by chakflying** [comment]
Also fixed a long standing issue where the ping line would get weirdly connected if the server was turned off for a while.

Screenshot 2024-01-27 012838

**Review by santosnepal — APPROVED** [review]

File: src/components/PingChart.vue
Comment (CommanderStorm) [inline src/components/PingChart.vue:13]: did you mean `type`?

```suggestion
                        type="button" class="dropdown-item" :class="{ active: chartPeriodHrs == key }"
```

**Review by CommanderStorm — COMMENTED** [review]
Overall, this looks very good to me. Massive props for all the work that went into this PR.

I have not found any defects, just a few maintainability/code style comments.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:352]: These labels are only for internal identification and not displayed to the user.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:461]: FYI, this is just a placeholder. UP datapoints are not displayed as a bar in the chart.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:262]: Honestly not sure, it was a bug that I encountered but I can't reproduce it right now.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:50]: The problem is that anything stored in localStorage always comes back as a string. The type conversions around this is a bit convoluted, but I have no interest in aligning everything as long as it works.

Comment (chakflying) [reply src/components/PingChart.vue:50]: We can do so in future PRs.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:510]: I'm not sure what you're asking.

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:223]: This value came from the dropdown and is not inputted by the user. I don't see why any validation would be needed.

File: src/components/PingChart.vue
Comment (CommanderStorm) [reply src/components/PingChart.vue:292]: Why `maintenance` is `undefined` instead of `0`.
I know that this is to match the backend, but I don't quite get why the functionality is like this in the backend.

File: src/components/PingChart.vue
Comment (CommanderStorm) [reply src/components/PingChart.vue:51]: Ok

Comment (CommanderStorm) [reply src/components/PingChart.vue:51]: good point

File: src/components/PingChart.vue
Comment (chakflying) [reply src/components/PingChart.vue:292]: maintenance is stored as json data in an `extras` column. It makes no sense to keep storing `maintenance: 0` as that would just waste space, and indeed it was implemented to not do that. That said, aligning the data format here is more just code style and should have no impact on the functionality.

**Review by CommanderStorm — APPROVED** [review]
From my point of view, this is mergable. (tested and this works as intended)
@louislam would you like to review this too?

**Comment by CommanderStorm** [comment]
> From my point of view, this is mergable. (tested and this works as intended)
> @louislam would you like to review this too?

I guess that is a no on an additional review.
=> merging with junior maintainer approval
(if you still have something, I will obv. provide a PR as previously done ^^)

**Pull request merged**
